File: forms/seasonal-affective-disorder-assessment/full-stack-with-loco-tera-htmx-alpine/tests/engine/sad_grader_test.rs
use seasonal_affective_disorder_assessment_tera_crate::engine::sad_grader::grade;
use seasonal_affective_disorder_assessment_tera_crate::engine::types::*;

/// A baseline data set with SPAQ items at 0 and PHQ-9 items at 0 (all answered).
fn answered_minimum() -> AssessmentData {
    AssessmentData {
        current_mood: CurrentMood {
            phq9: Phq9Items {
                q1: Some(0),
                q2: Some(0),
                q3: Some(0),
                q4: Some(0),
                q5: Some(0),
                q6: Some(0),
                q7: Some(0),
                q8: Some(0),
                q9: Some(0),
            },
            difficulty_level: "not-difficult".to_string(),
        },
        sleep_energy: SleepEnergy {
            spaq: SpaqSleep {
                sleep_length: Some(0),
                energy_level: Some(0),
            },
            ..Default::default()
        },
        appetite_weight: AppetiteWeight {
            spaq: SpaqAppetite {
                appetite: Some(0),
                weight: Some(0),
            },
            ..Default::default()
        },
        social_occupational: SocialOccupational {
            spaq: SpaqSocial {
                mood: Some(0),
                social_activity: Some(0),
            },
            ..Default::default()
        },
        ..Default::default()
    }
}

#[test]
fn empty_assessment_is_no_sad() {
    let data = AssessmentData::default();
    let result = grade(&data);
    assert_eq!(result.spaq_score, 0);
    assert_eq!(result.phq9_score, 0);
    assert_eq!(result.spaq_band, "no-sad");
    assert_eq!(result.phq9_band, "minimal");
    assert_eq!(result.combined_severity, "no-sad");
    assert!(result.fired_rules.is_empty(), "no items answered => no fired rules");
}

#[test]
fn baseline_zero_scores_are_no_sad() {
    let data = answered_minimum();
    let result = grade(&data);
    assert_eq!(result.spaq_score, 0);
    assert_eq!(result.phq9_score, 0);
    assert_eq!(result.spaq_band, "no-sad");
    assert_eq!(result.phq9_band, "minimal");
    assert_eq!(result.combined_severity, "no-sad");
    // 6 SPAQ + 9 PHQ-9 items answered.
    assert_eq!(result.fired_rules.len(), 15);
}

#[test]
fn sad_likely_band_classification() {
    let mut data = answered_minimum();
    // SPAQ total = 11+ → "sad-likely".
    data.sleep_energy.spaq.sleep_length = Some(2);
    data.sleep_energy.spaq.energy_level = Some(2);
    data.appetite_weight.spaq.appetite = Some(2);
    data.appetite_weight.spaq.weight = Some(2);
    data.social_occupational.spaq.mood = Some(2);
    data.social_occupational.spaq.social_activity = Some(2);
    let result = grade(&data);
    assert_eq!(result.spaq_score, 12);
    assert_eq!(result.spaq_band, "sad-likely");
}

#[test]
fn subsyndromal_band_classification() {
    let mut data = answered_minimum();
    // SPAQ total = 9 → "subsyndromal".
    data.sleep_energy.spaq.sleep_length = Some(2);
    data.sleep_energy.spaq.energy_level = Some(2);
    data.appetite_weight.spaq.appetite = Some(2);
    data.appetite_weight.spaq.weight = Some(2);
    data.social_occupational.spaq.mood = Some(1);
    data.social_occupational.spaq.social_activity = Some(0);
    let result = grade(&data);
    assert_eq!(result.spaq_score, 9);
    assert_eq!(result.spaq_band, "subsyndromal");
}

#[test]
fn phq9_severe_band_and_critical_combined() {
    let mut data = answered_minimum();
    data.current_mood.phq9.q1 = Some(3);
    data.current_mood.phq9.q2 = Some(3);
    data.current_mood.phq9.q3 = Some(3);
    data.current_mood.phq9.q4 = Some(3);
    data.current_mood.phq9.q5 = Some(3);
    data.current_mood.phq9.q6 = Some(3);
    data.current_mood.phq9.q7 = Some(3);
    data.current_mood.phq9.q8 = Some(2);
    data.current_mood.phq9.q9 = Some(0); // q9=0 isolates phq9_score path
    let result = grade(&data);
    assert_eq!(result.phq9_score, 23);
    assert_eq!(result.phq9_band, "severe");
    // PHQ-9 total >= 20 OR PHQ-9 band 'severe' alone → "critical" by the
    // first rule (phq9_score >= 20).
    assert_eq!(result.combined_severity, "critical");
}

#[test]
fn suicidal_ideation_yields_critical_severity() {
    let mut data = answered_minimum();
    data.risk_assessment.suicidal_ideation = "yes".to_string();
    let result = grade(&data);
    assert_eq!(result.combined_severity, "critical");
    assert!(result.additional_flags.iter().any(|f| f.id == "FLAG-RISK-001"));
}

#[test]
fn phq9_q9_positive_yields_critical_severity() {
    let mut data = answered_minimum();
    data.current_mood.phq9.q9 = Some(1);
    let result = grade(&data);
    assert_eq!(result.combined_severity, "critical");
    assert!(result.additional_flags.iter().any(|f| f.id == "FLAG-PHQ9-Q9"));
}

#[test]
fn mild_severity_path_via_phq9_mild() {
    let mut data = answered_minimum();
    // PHQ-9 total 5 → "mild".
    data.current_mood.phq9.q1 = Some(2);
    data.current_mood.phq9.q2 = Some(2);
    data.current_mood.phq9.q3 = Some(1);
    let result = grade(&data);
    assert_eq!(result.phq9_score, 5);
    assert_eq!(result.phq9_band, "mild");
    assert_eq!(result.combined_severity, "mild");
}

#[test]
fn moderate_severity_path_via_phq9_moderate() {
    let mut data = answered_minimum();
    // PHQ-9 total 10 → "moderate".
    data.current_mood.phq9.q1 = Some(2);
    data.current_mood.phq9.q2 = Some(2);
    data.current_mood.phq9.q3 = Some(2);
    data.current_mood.phq9.q4 = Some(2);
    data.current_mood.phq9.q5 = Some(2);
    let result = grade(&data);
    assert_eq!(result.phq9_score, 10);
    assert_eq!(result.phq9_band, "moderate");
    assert_eq!(result.combined_severity, "moderate");
}

#[test]
fn rule_ids_are_unique() {
    let data = answered_minimum();
    let result = grade(&data);
    let mut ids: Vec<&str> = result.fired_rules.iter().map(|r| r.id.as_str()).collect();
    ids.sort();
    let len_before = ids.len();
    ids.dedup();
    assert_eq!(ids.len(), len_before, "rule IDs must be unique");
}

#[test]
fn fired_rules_carry_score_value() {
    let mut data = answered_minimum();
    data.sleep_energy.spaq.sleep_length = Some(3);
    let result = grade(&data);
    let r = result
        .fired_rules
        .iter()
        .find(|r| r.id == "SPAQ-001")
        .expect("SPAQ-001 must fire");
    assert_eq!(r.score, 3);
}

File: forms/seasonal-affective-disorder-assessment/front-end-with-svelte/src/lib/engine/sad-grader.test.ts
import { describe, it, expect } from 'vitest';
import { calculateSadGrade } from './sad-grader';
import { detectAdditionalFlags } from './flagged-issues';
import { spaqItems, phq9Items, classifySpaq, classifyPhq9 } from './sad-rules';
import type { AssessmentData } from './types';

/** A fully-blank assessment, built inline so the engine tests have no
 *  dependency on the SvelteKit store (which imports `$app/environment`). */
function blank(): AssessmentData {
	return {
		demographics: {
			firstName: '',
			lastName: '',
			dateOfBirth: '',
			sex: '',
			latitude: '',
			country: '',
			yearsAtCurrentLatitude: null
		},
		seasonalPatternHistory: {
			symptomsRecurAnnually: '',
			worstMonths: '',
			bestMonths: '',
			yearsAffected: null,
			familyHistorySad: '',
			firstOnsetAge: ''
		},
		currentMood: {
			phq9: { q1: null, q2: null, q3: null, q4: null, q5: null, q6: null, q7: null, q8: null, q9: null },
			difficultyLevel: ''
		},
		sleepEnergy: {
			spaq: { sleepLength: null, energyLevel: null },
			hoursSleptWinter: null,
			hoursSleptSummer: null,
			hypersomnia: '',
			morningFatigue: '',
			energyNotes: ''
		},
		appetiteWeight: {
			spaq: { appetite: null, weight: null },
			carbohydrateCraving: '',
			winterWeightChangeKg: null,
			eatingPatternChanges: ''
		},
		socialOccupational: {
			spaq: { mood: null, socialActivity: null },
			workImpaired: '',
			relationshipsImpaired: '',
			socialWithdrawal: '',
			occupationalNotes: ''
		},
		lightExposure: {
			dailyOutdoorMinutes: null,
			workIndoors: '',
			curtainsClosedDaytime: '',
			sunriseExposure: '',
			usesLightTherapyBox: '',
			lightTherapyDetails: '',
			lightTherapyAccess: ''
		},
		previousTreatments: {
			antidepressants: '',
			antidepressantDetails: '',
			psychotherapy: '',
			psychotherapyDetails: '',
			lightTherapyHistory: '',
			lightTherapyHistoryDetails: '',
			currentTreatment: '',
			currentTreatmentDetails: ''
		},
		riskAssessment: {
			suicidalIdeation: '',
			suicidalIntent: '',
			suicidalPlan: '',
			selfHarm: '',
			selfHarmDetails: '',
			previousAttempt: '',
			protectiveFactors: '',
			safetyPlanInPlace: ''
		},
		treatmentPlan: {
			planLightTherapy: '',
			planAntidepressant: '',
			planPsychotherapy: '',
			planLifestyle: '',
			planCrisisReferral: '',
			followUpInterval: '',
			clinicianNotes: ''
		}
	};
}

describe('SAD grading engine', () => {
	it('returns no-sad for a fully blank assessment', () => {
		const r = calculateSadGrade(blank());
		expect(r.spaqScore).toBe(0);
		expect(r.spaqBand).toBe('no-sad');
		expect(r.phq9Score).toBe(0);
		expect(r.phq9Band).toBe('minimal');
		expect(r.combinedSeverity).toBe('no-sad');
		expect(r.firedRules).toHaveLength(0);
	});

	it('sums the six SPAQ items into the 0-24 GSS', () => {
		const d = blank();
		d.sleepEnergy.spaq.sleepLength = 4;
		d.sleepEnergy.spaq.energyLevel = 4;
		d.socialOccupational.spaq.mood = 4;
		d.socialOccupational.spaq.socialActivity = 4;
		d.appetiteWeight.spaq.appetite = 2;
		d.appetiteWeight.spaq.weight = 2;
		const r = calculateSadGrade(d);
		expect(r.spaqScore).toBe(20);
		expect(r.spaqBand).toBe('sad-likely');
	});

	it('sums the nine PHQ-9 items into the 0-27 score', () => {
		const d = blank();
		d.currentMood.phq9.q1 = 2;
		d.currentMood.phq9.q2 = 2;
		d.currentMood.phq9.q3 = 1;
		const r = calculateSadGrade(d);
		expect(r.phq9Score).toBe(5);
		expect(r.phq9Band).toBe('mild');
	});

	it('classifies mild from subsyndromal SPAQ', () => {
		const d = blank();
		d.sleepEnergy.spaq.sleepLength = 3;
		d.sleepEnergy.spaq.energyLevel = 3;
		d.socialOccupational.spaq.mood = 3;
		const r = calculateSadGrade(d);
		expect(r.spaqScore).toBe(9);
		expect(r.spaqBand).toBe('subsyndromal');
		expect(r.combinedSeverity).toBe('mild');
	});

	it('classifies moderate from SAD-likely SPAQ', () => {
		const d = blank();
		for (const item of spaqItems) {
			const sub = d[item.section] as unknown as Record<string, Record<string, number | null>>;
			sub[item.subsection][item.field] = 2;
		}
		const r = calculateSadGrade(d);
		expect(r.spaqScore).toBe(12);
		expect(r.spaqBand).toBe('sad-likely');
		expect(r.combinedSeverity).toBe('moderate');
	});

	it('escalates to critical on PHQ-9 item 9', () => {
		const d = blank();
		d.currentMood.phq9.q9 = 1;
		const r = calculateSadGrade(d);
		expect(r.combinedSeverity).toBe('critical');
	});

	it('escalates to critical on suicidal ideation', () => {
		const d = blank();
		d.riskAssessment.suicidalIdeation = 'yes';
		const r = calculateSadGrade(d);
		expect(r.combinedSeverity).toBe('critical');
	});

	it('escalates to critical on PHQ-9 total >= 20', () => {
		const d = blank();
		d.currentMood.phq9.q1 = 3;
		d.currentMood.phq9.q2 = 3;
		d.currentMood.phq9.q3 = 3;
		d.currentMood.phq9.q4 = 3;
		d.currentMood.phq9.q5 = 3;
		d.currentMood.phq9.q6 = 3;
		d.currentMood.phq9.q7 = 2;
		const r = calculateSadGrade(d);
		expect(r.phq9Score).toBe(20);
		expect(r.combinedSeverity).toBe('critical');
	});

	it('has unique SPAQ + PHQ item ids', () => {
		const ids = [...spaqItems.map((i) => i.id), ...phq9Items.map((i) => i.id)];
		expect(new Set(ids).size).toBe(ids.length);
	});
});

describe('SPAQ / PHQ-9 band classifiers', () => {
	it('classifies SPAQ bands at the boundaries', () => {
		expect(classifySpaq(7)).toBe('no-sad');
		expect(classifySpaq(8)).toBe('subsyndromal');
		expect(classifySpaq(10)).toBe('subsyndromal');
		expect(classifySpaq(11)).toBe('sad-likely');
	});
	it('classifies PHQ-9 bands at the boundaries', () => {
		expect(classifyPhq9(4)).toBe('minimal');
		expect(classifyPhq9(9)).toBe('mild');
		expect(classifyPhq9(14)).toBe('moderate');
		expect(classifyPhq9(19)).toBe('moderately-severe');
		expect(classifyPhq9(20)).toBe('severe');
	});
});

describe('SAD flagged issues', () => {
	it('returns no flags for a blank assessment', () => {
		const d = blank();
		const flags = detectAdditionalFlags(d, {
			phq9Score: 0,
			spaqBand: 'no-sad',
			combinedSeverity: 'no-sad'
		});
		expect(flags).toHaveLength(0);
	});

	it('flags active suicidal ideation as high priority', () => {
		const d = blank();
		d.riskAssessment.suicidalIdeation = 'yes';
		const flags = detectAdditionalFlags(d, {
			phq9Score: 0,
			spaqBand: 'no-sad',
			combinedSeverity: 'critical'
		});
		expect(flags.some((f) => f.id === 'FLAG-RISK-001' && f.priority === 'high')).toBe(true);
	});

	it('flags low daily outdoor light exposure', () => {
		const d = blank();
		d.lightExposure.dailyOutdoorMinutes = 10;
		const flags = detectAdditionalFlags(d, {
			phq9Score: 0,
			spaqBand: 'no-sad',
			combinedSeverity: 'no-sad'
		});
		expect(flags.some((f) => f.id === 'FLAG-LIFE-001')).toBe(true);
	});

	it('sorts flags by priority (high first)', () => {
		const d = blank();
		d.riskAssessment.suicidalIdeation = 'yes';
		d.lightExposure.workIndoors = 'yes';
		const flags = detectAdditionalFlags(d, {
			phq9Score: 0,
			spaqBand: 'no-sad',
			combinedSeverity: 'critical'
		});
		const order = { high: 0, medium: 1, low: 2 };
		const ps = flags.map((f) => f.priority);
		expect(ps).toEqual([...ps].sort((a, b) => order[a] - order[b]));
	});
});
